script: Do not set up an IPC route for every image load

mrobinson opened the pull request:
Instead of setting up a route for every image load in the DOM / Layout,
route all incoming image cache responses through the `ScriptThread`.
This avoids creating a set of file descriptors for every image that is
loaded.

This change requires having the `ImageCache` track the `PipelineId` of
the original the listener so that the `ScriptThread` can route it
properly to the correct `Window`.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes are part of #33672.
- [x] These changes are covered by existing WPT tests as they should not change page-observable behavior.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
script: Do not set up an IPC route for every image load

Instead of setting up a route for every image load in the DOM / Layout,
route all incoming image cache responses through the `ScriptThread`.
This avoids creating a set of file descriptor for every image that is
loaded.

This change requires having the `ImageCache` track the `PipelineId` of
the original the listener so that the `ScriptThread` can route it
properly to the correct `Window`.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

jdm commented on components/script/dom/htmlvideoelement.rs:
```diff
@@ -177,28 +177,58 @@ impl HTMLVideoElement {
+                    .as_global_scope()
+                    .task_manager()
+                    .networking_task_source()
+                    .queue(task!(process_image_response: move || {
```
Since we're already invoking the callback in the script thread, using a task here just adds latency. We should be able to skip the task and run the actual callback code directly.

mrobinson replied:
I gave this a shot (the version of this that failed the merge queue) and it seems that it caused some problems. Looking over the HTML image element code it seems that various parts of the loading algorithm run in tasks already. I think that the code was depending in some way on the order of events being enforced by the task queue. It's not totally clear to me if it's possible to disentangle just these calls from the queue. @jdm, given that the HTML image element was already queuing this response handling as a task, is it okay to maintain this code for now?

I think that the video element loader doesn't present the same problem and I believe I can make the change you requested there.

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -380,15 +374,62 @@ impl HTMLImageElement {
+                .as_global_scope()
+                .task_manager()
+                .networking_task_source()
+                .queue(task!(process_image_response: move || {
```
Same point about skipping the task enqueuing step here.

jdm commented on components/script/dom/window.rs:
```diff
@@ -555,6 +569,26 @@ impl Window {
+        };
+
+        for callback in callbacks.get() {
+            callback(self, response.clone());
```
If we end up removing the task queueing, we'll have an outstanding mutable borrow here and callbacks that contain `CanGc::note()`.

mrobinson replied:
I have gone ahead and done a `std::mem::take` here for the duration of the method.

jdm commented on components/script/dom/window.rs:
```diff
@@ -311,6 +308,13 @@ pub(crate) struct Window {
+    /// layout, they also add an etry to [`Self::pending_layout_images`].
+    #[no_trace]
+    #[ignore_malloc_size_of = "A boxed Fn is very hard to measure"]
+    pending_image_callbacks: DomRefCell<HashMap<PendingImageId, Vec<PendingImageCallback>>>,
```
Can we pushing the ignore_malloc_size_of onto the actual type that doesn't implement it somehow? It would be useful to see if this hashmap ends up growing unexpectedly, but we can't observe that right now.

mrobinson replied:
Yep, this is done now.

jdm reviewed (approved):
This looks reasonable!

gterzian commented on components/script/dom/window.rs:
```diff
@@ -311,6 +308,13 @@ pub(crate) struct Window {
+    /// When an element triggers an image load or starts watching an image load from the
+    /// `ImageCache` it adds an entry to this list. When those loads are triggered from
+    /// layout, they also add an etry to [`Self::pending_layout_images`].
+    #[no_trace]
```
I don't like the `&Window` in the callback, and the fact that this is not traced. I think it probably works because the `&Window` corresponds to the `&self` that registers and invokes the callback, but I don't like that fact that this is yet another mechanism to understand(on top of what we do with rooting elsewhere).

Also this mechanism is tied to a window, and while this may fit the requirements of image loading it would not fit something that does not always have a window available, which is everything that currently uses an IPC route but may be found either in a worker or window. 

For an example of what I do like: the [callback](https://github.com/servo/servo/blob/e74539404467cc31c03f52a419371258235a1d69/components/script/dom/promisenativehandler.rs#L17) mechanism for native promise handlers because:
- it can be used for anything that needs to work with promises in any event-loop context. 
- it leverages the existing rooting mechanisms. 

So I would like us to find a solution for the problem of too many IPC routes in a similar way. 

The current use of IPC routes might be expensive, but it has a great advantage in terms of ease of use and consistency across various use cases. I would not like to see this replaced by many different ad-hoc solutions, like this one, or the `FetchThead`, and one for each other use case.

gterzian reviewed (commented):
I am aware of the recent awareness around IPC routes being too expensive, I think we are discussing it over at https://github.com/servo/servo/issues/34918 and I would rather have us design a mechanism that works across all use cases, and then implement it, as opposed to what is proposed in this PR and what is already done with the `FetchThread`, and what would have to be done in other places that also use or will use an IPC route, which amounts to a multitude of different solutions. 

I also prefer to first discuss these things outside of a PR, because it is much more work to read the code in a PR and then go backwards to a design, and the fact that a PR presents a working solution makes it in my opinion also harder to consider alternatives.

mrobinson commented:
> I also prefer to first discuss these things outside of a PR, because it is much more work to read the code in a PR and then go backwards to a design, and the fact that a PR presents a working solution makes it in my opinion also harder to consider alternatives.

If a better alternative is proposed, I do not think it will be harder to consider. I will be happy to review it myself.  I agree that we should look for a more general solution to the IPC routing issues. That said, my primary priority is to fix the crash in Servo and this is a viable solution to do that. It may take a long time to devise an IPC routing system that does not leak file descriptors and in the meantime I think we need to make sure that it is possible to run WPT tests on macOS systems.

This solution is unlike the `FetchThread` one. This time I have made sure to integrate the messaging into the `ScriptThread` event loop which is something you objected to in the `FetchThread` change. I still have pending to remove the `FetchThread` -- something that I hope to to do in the near future. 

> I also prefer to first discuss these things outside of a PR, because it is much more work to read the code in a PR and then go backwards to a design, and the fact that a PR presents a working solution makes it in my opinion also harder to consider alternatives.

I have removed the `&Window` argument to the callback. I think this addresses your concerns. It's true that only elements in `Window`s currently trigger image loading, but now it should be possible outside the `Window` context as well. Thanks for pointing that out. It's true that this is currently tied to the `Window` now, but it could easily be stored on the `GlobaScope` if it's necessary in other contexts. It is currently on `Window` at the moment simply because it isn't necessary in other types of `GlobalScope`s.

> The current use of IPC routes might be expensive, but it has a great advantage in terms of ease of use and consistency across various use cases. I would not like to see this replaced by many different ad-hoc solutions, like this one, or the FetchThead, and one for each other use case

I understand that the IPC route approach is very clean, but I think that a mechanism that exhausts system resources is a non-starter for production software. I'm happy to make the change to switch to a cleaner system when it exists. That's what I was proposing on Zulip. Here's my proposal:

1. Fix the high priority overuse of IPC routing such that Servo does not crash.
2. Come up with a more generic system for handling this type of situation.
3. Convert everything to use the new system.

mrobinson commented:
> The current use of IPC routes might be expensive, but it has a great advantage in terms of ease of use and consistency across various use cases. I would not like to see this replaced by many different ad-hoc solutions, like this one, or the FetchThead, and one for each other use case

This may be conceptually a little more complicated, but only if you are comfortable with how we use `ROUTER` in Servo which honestly I really only understood within the last year. Despite the increase in complication, I think it's interesting to point out that this is a decrease in the amount of code.

gterzian commented:
> It may take a long time to devise an IPC routing system

Makes sense. 

> This time I have made sure to integrate the messaging

Yes I prefer this iteration to the `FetchThread`, but I don't like the fact that both end-up as a separate solution in the codebase.

> Here's my proposal
Sounds good.  I'm ok with an iterative approach(although it might make it somewhat harder to come to a general solution, because it's one thing to replace one pattern with another, and it's another to merge several patterns into a new one).

mrobinson commented:
I think I will go ahead and land this one. I was not able to remove the task queuing from `HTMLImageElement`, but it was a pre-existing design of the code there. It seems that the current code is using the task queue to ensure the ordering of operations when processing changes to the image element attributes. Removing this requires understanding the specification to see if the new design would still be compliant. This is a much larger task which I think might be better addressed in later work.
